forem/forem
# Add service generator

## What type of PR is this? (check all applicable)


## Description

I'm currently a bit blocked on profile work, waiting for 2 PRs to get approved and merged. So in the meantime I did some work on a service generator to standardize our service objects a bit more.

### Examples:

**Generate a non-namespaced service without arguments**

`$ rails generate service DoTheThing`

```ruby
# app/services/do_the_thing.rb
class DoTheThing
  def self.call
    new.call
  end

  def call
  end
end
```

```ruby
# spec/services/do_the_thing_spec.rb
require "rails_helper"

RSpec.describe DoTheThing, type: :service do
  pending "add some examples to (or delete) #{__FILE__}"
end
```

**Generate a non-namespaced service with arguments:**

`$ rails generate service DoTheThing arg1 arg2`

```ruby
# app/services/do_the_thing.rb
class DoTheThing
  def self.call(arg1, arg2)
    new(arg1, arg2).call
  end

  def initialize(arg1, arg2)
    @arg1 = arg1
    @arg2 = arg2
  end

  def call
  end
end
```

The generated spec is the same as above.

**Generate a namespaced service with arguments**

`$ rails generate service things/dothem arg1 arg2`

```ruby
# app/services/things/do_them.rb
class Things::DoThem
  def self.call(arg1, arg2)
    new(arg1, arg2).call
  end

  def initialize(arg1, arg2)
    @arg1 = arg1
    @arg2 = arg2
  end

  def call
  end
end
```

```ruby
# spec/services/things/do_them_spec.rb
require "rails_helper"

RSpec.describe Things::DoThem, type: :service do
  pending "add some examples to (or delete) #{__FILE__}"
end
```

## Related Tickets & Documents

n/a

## QA Instructions, Screenshots, Recordings

Try the different examples above and verify that the generated code matches your expectations.

## Added tests?

- [X] No, and this is why: there's surprisingly little documentation around testing generators with RSpec. The examples I found all use an external gem for this. I'm not sure I want to add this dependency just yet and we are also not testing the other generator. As this is a convenience tool for developers and not production code I think this is ok until I can evaluate testing strategies during the next cooldown.

## Added to documentation?
